core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -391,8 +454,21 @@ public void requestConnection() {
         scheduleNextConnection();
         break;
       case TRANSIENT_FAILURE:
+        if (scheduleConnectionTask != null) {
+          break; // let the already scheduled task do its job
+        }
         addressIndex.increment();
-        requestConnection();
+        if (!serializingRetries) {
+          requestConnection();
+        } else {
+          if (!addressIndex.isValid()) {
+            scheduleBackoff();
+          } else {
+            subchannelData.subchannel.shutdown(); // shutdown the previous subchannel
```

Comment:
Don't we need to shutdown the subchannel as soon as it goes TF? Otherwise while we try other subchannels this one will start reconnecting when its backoff expires.

Reply:
Good point

Reply:
I think the shutdown needs to be in `processSubchannelState()`, as that method could have already called `addressIndex.increment()`. But we also have the problem that when you shutdown the subchannel you need to remove it from `subchannels` (missing in the recent addition), which will then need other tweaks.